pyro/ops/stats.py

```diff
@@ -0,0 +1,208 @@
+from __future__ import absolute_import, division, print_function
+
+import torch
+
+
+def _compute_chain_variance_stats(input):
+    # compute within-chain variance and variance estimator
+    # input has shape N x C x sample_shape
+    N = input.size(0)
+    chain_mean = input.mean(dim=0)
+    var_between = chain_mean.var(dim=0)
+
+    chain_var = input.var(dim=0)
+    var_within = chain_var.mean(dim=0)
+    var_estimator = (N - 1) / N * var_within + var_between
+    return var_within, var_estimator
+
+
+def gelman_rubin(input, sample_dim=0, chain_dim=1):
+    """
+    Computes R-hat over chains of samples. It is required that
+    `input.size(sample_dim) >= 2` and `input.size(chain_dim) >= 2`.
+
+    :param torch.Tensor input: the input tensor.
+    :param int sample_dim: the sample dimension.
+    :param int chain_dim: the chain dimension.
+    :returns torch.Tensor: R-hat of `input`.
+    """
+    assert input.dim() >= 2
+    assert input.size(sample_dim) >= 2
+    assert input.size(chain_dim) >= 2
+    # change input.shape to 1 x 1 x input.shape
+    # then transpose sample_dim with 0, chain_dim with 1
+    sample_dim = input.dim() + sample_dim if sample_dim < 0 else sample_dim
+    chain_dim = input.dim() + chain_dim if chain_dim < 0 else chain_dim
+    assert chain_dim != sample_dim
+    input = input.reshape((1, 1) + input.shape)
+    input = input.transpose(0, sample_dim + 2).transpose(1, chain_dim + 2)
+
+    var_within, var_estimator = _compute_chain_variance_stats(input)
+    rhat = (var_estimator / var_within).sqrt()
+    return rhat.squeeze(max(sample_dim, chain_dim)).squeeze(min(sample_dim, chain_dim))
+
+
+def split_gelman_rubin(input, sample_dim=0, chain_dim=1):
+    """
+    Computes R-hat over chains of samples. It is required that
+    `input.size(sample_dim) >= 4`.
+
+    :param torch.Tensor input: the input tensor.
+    :param int sample_dim: the sample dimension.
+    :param int chain_dim: the chain dimension.
+    :returns torch.Tensor: split R-hat of `input`.
+    """
+    assert input.dim() >= 2
+    assert input.size(sample_dim) >= 4
+    # change input.shape to 1 x 1 x input.shape
+    # then transpose sample_dim with 0, chain_dim with 1
+    sample_dim = input.dim() + sample_dim if sample_dim < 0 else sample_dim
+    chain_dim = input.dim() + chain_dim if chain_dim < 0 else chain_dim
+    assert chain_dim != sample_dim
+    input = input.reshape((1, 1) + input.shape)
+    input = input.transpose(0, sample_dim + 2).transpose(1, chain_dim + 2)
+
+    N_half = input.size(0) // 2
+    new_input = torch.stack([input[:N_half], input[-N_half:]], dim=1)
+    new_input = new_input.reshape((N_half, -1) + input.shape[2:])
+    split_rhat = gelman_rubin(new_input)
+    return split_rhat.squeeze(max(sample_dim, chain_dim)).squeeze(min(sample_dim, chain_dim))
+
+
+def _fft_next_good_size(N):
```

Comment:
Are you able to run these tests using the night PyTorch build. I get `RuntimeError: fft: ATen not compiled with MKL support`, so this will likely also fail when we merge into `pytorch-1.0` branch.

Reply:
According to [rfft docs](https://pytorch.org/docs/stable/torch.html#torch.rfft), I think you need to install pytorch with MKL.

Reply:
The pip wheels do not ship with MKL unfortunately, but the conda package does, which also implies that we need to run most of these tests locally since CI will end up skipping.